Create Rule “do-you-optimize-tinacms-project-for-clarity-performance-and-reliable-builds/rule” (#9076)
SSWConsulting/SSW.Rules.Content: rules/do-you-optimize-tinacms-project/rule.md

---
type: rule
title: Do you optimize your TinaCMS project for clarity, performance, and reliable builds?
seoDescription: Learn effective strategies to structure and optimize your
  TinaCMS project for clarity, performance, and successful builds.
uri: do-you-optimize-tinacms-project
authors:
  - title: "Gilles Pothieu"
    url: https://www.ssw.com.au/people/gilles-pothieu/
created: 2024-08-28T16:15:00.000Z
guid: 3c59f56f-34b0-47a7-902a-718eb340d2e5
---
Structuring and optimizing your TinaCMS project is essential to achieve clarity, enhance performance, and prevent build failures. Poorly optimized projects can lead to slow site performance, increased server load, and even failed builds due to excessive or inefficient data requests.

Let’s explore how to structure your project effectively and apply best practices to boost performance both in runtime and during the build process.

<!--endintro-->

## 1. Structuring your TinaCMS Architecture

When working with large datasets or generating multiple subcomponents, following best practices is crucial to maintain performance and clarity.

### ❌ Bad practices

> * **Making individual requests with Tina client for each Subcomponent**
> 
>   * This method can overwhelm the build process, particularly when dealing with a large number of subcomponents.

Reviewer: ```suggestion
```
this was a point in time issue with Tina and has been fixed

Author: ✅ Done

* **Using deeply nested schemas with nested references**

  * Complex and deeply nested schemas increase the complexity of the project, making it harder to manage and more prone to build failures.
  * They can also lead to inefficient data fetching, further slowing down both runtime and build processes.

### ✅ Good practices

* **Making a single request at a top-level server component and passing props down**

  * Data fetched at the top level is cached by default, which enhances performance by reducing redundant API calls.
  * Subcomponents can directly access the necessary data, eliminating the need for repeated requests.

::: good
![Figure: Good example - Single request at the top-level server and passing props down](2024-08-28_16-21-56.png)

> :::

Reviewer: not sure about this as it suggests that god objects are the best approach, which leads to prop-drilling - which is an anti-pattern in react

Author: ✅ I changed that part for React context example and explanation

### ❌ Bad practices

* **Using client-side requests instead of cached values from server-side components**

  * This approach bypasses pre-fetched, cached data, increasing server load and slowing down the application.
  * It negates the benefits of static site generation, where data should be ready and waiting.

### Common practices

* **Utilizing Next.js caching via fetch** 

  * Next.js’s fetch API is designed to cache by default during static generation and server-side rendering.

### ✅ Good practices

* **Utilizing caching at the highest level of the application**

  * By caching data at the top level, you ensure that it is readily available throughout your application, reducing redundant data fetching and improving overall performance.
  * This approach allows for more efficient data management and quicker response times.
* **Using dedicated caching packages**

  * [**Redis**:](https://redis.io/fr/solutions/cas-dutilisation/cache/) An in-memory data structure store, Redis is fast and versatile, making it ideal for caching data in memory. 
  * [**Node-cache**:](https://www.npmjs.com/package/node-cache) A lightweight, in-memory caching solution for Node.js applications.

> 

Reviewer 2: still unsure about this section. I feel we should mention to statically generate the content i.e. calling it only on build. Or using a fetching library like react query that can cache frequent requests.

Reviewer: agree - it also focuses on nextjs and react as opposed to TinaCMS

Author: I changed this part, more focus on Tina, not on next/react.

> ## 3. Preventing Build Crashes
> 
> To ensure smooth and reliable builds, it’s important to follow best practices that prevent excessive server load and manage data efficiently.

Reviewer: ```suggestion
## 3. Improving Build performance

To ensure smooth and reliable builds, it’s important to follow best practices that prevent excessive server load and manage data efficiently.
```

### ✅ Best practices

> * **Avoid isolating Tina client calls in multiple subcomponents**
> 
>   * Centralize data fetching to avoid overwhelming the build process with excessive API requests.
> 
>      E.g. If your application has a footer and a header that fetch data every time you navigate to a new page, you can improve performance by caching this data from a top-level call.

Reviewer: ```suggestion
```
this will be resolved with build caching

Author: ✅ Done

> * **Cache large datasets at a top level**
> 
>   * By caching large datasets at a top level, you reduce the number of data fetches during the build process, lowering the risk of overwhelming the system.
> 
>     E.g. If your application has a list of recipes that fetch data each time you view a recipe, caching all the data at the top level instead of fetching it each time would reduce API calls during the build process.

Reviewer: ```suggestion
```
god object comment

Author: ✅ Done

* **Avoid using deeply nested references**

  * Simplifying your data structure and avoiding deeply nested references can help ensure smoother and faster builds.

* **Write custom Tina queries**

>   * You can improve build generation and prevent build crash by creating your own Tina GraphQL queries.  

Reviewer: ```suggestion
  * You can improve data retrieval performance by creating your own GraphQL queries 
```

> * **Write custom Tina queries**
> 
>   * You can improve build generation and prevent build crash by creating your own Tina GraphQL queries.  
> 
>     Some of the auto-generated Tina queries are not optimized, such as those with nested objects containing redundant data. For example, recipes that include an ingredients object, which in turn includes the same recipes again. Creating custom queries can reduce the size of objects and improve performance.

Reviewer: ```suggestion
* **Write custom GraphQL queries**
  * You can improve data retreival by creating your own GraphQL queries
    Auto-generated GraphQL queries are not optimized, as a result, they may include nested objects with redundant data. For example, recipes that include an ingredients object, which in turn includes the same recipes again. Creating custom queries can reduce the size of objects and improve performance
```

Author: ✅ Done

---
type: rule
title: Do you optimize your TinaCMS project for clarity, performance, and reliable builds?
seoDescription: Learn effective strategies to structure and optimize your
    TinaCMS project for clarity, performance, and successful builds.
uri: do-you-optimize-tinacms-project
authors:
    - title: 'Gilles Pothieu'
      url: https://www.ssw.com.au/people/gilles-pothieu/
created: 2024-08-28T16:15:00.000Z
guid: 3c59f56f-34b0-47a7-902a-718eb340d2e5
---

> Structuring and optimizing your TinaCMS project is essential to achieve clarity, enhance performance, and prevent build failures. Poorly optimized projects can lead to slow site performance, increased server load, and even failed builds due to excessive or inefficient data requests.

Reviewer 2: ```suggestion
Setting up and optimizing your TinaCMS project properly is key to maintaining clear organization, maximizing performance, and preventing build issues. Without optimization, your site may suffer from slow load times, increased server strain, and failed builds due to inefficient data handling.
```

Not sure I love the intro sentence but can't come up with anything better than it 😅

  * Data fetched at the top level can be stored in a React Context or a global state management solution (e.g., [Redux](https://redux.js.org/)). This allows all components to access the data without the need to pass props manually
  * This approach ensures better scalability, as subcomponents can access the necessary data directly from the context or store, eliminating redundant API calls and avoiding prop drilling

::: good
```js
export default async function Home({ params }: HomePageProps) {
  const location = params.location;
  
  // Appel à l'API ou client GraphQL
  const websiteProps = await client.queries.website({
    relativePath: `${location}/website.md`,
  });

  // Imaginons que websiteProps contient ces données
  const { conferencesData, footerData, speakers } = websiteProps.data;

  return (
      <ConferenceContext.Provider value={conferencesData}>
        <FooterContext.Provider value={footerData}>
          <PageTransition>
            <HomeComponent speakers={speakers}/>
          </PageTransition>
        </FooterContext.Provider>
      </ConferenceContext.Provider>
  );
}

> // Génération des paramètres statiques pour SSG

Reviewer 3: Please change this to English, or remove the comment

Author: ✅ Done
